### PR TITLE
enhancement/global-state

### PR DESCRIPTION
**Pull request by aminsaedi**

### CODE REVIEW

File: src/store/doc.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import { createSlice, PayloadAction } from "@reduxjs/toolkit";
+
+type DocType = {
+  activeTabId: string;
```
Comment (pmvald) [inline src/store/doc.ts:4]: we need another layer for tabId, so the doc structure is something like

```
{
activeTabIndices: {
   [tabId]: string (activeTabIndex)
}
}
```

File: docs-src/pgweb/01-topic.mdx [hunk 2]
```diff
@@ -10,13 +10,13 @@ Notice **Two** with link [link](https://angular.io/)
...
 <Tabs>
-<Tab label="First tab">
+<Tab id="python" label="Python">
 
 Inside Tab 1: **Two** with link [link](https://angular.io/)
 
 </Tab>
 
-<Tab label="Second tab">
+<Tab id="javaScript" label="Java-Script">
```
Comment (pmvald) [inline docs-src/pgweb/01-topic.mdx:19]: no need for this change

**Review by pmvald — CHANGES_REQUESTED** [review]

**Comment by aminsaedi** [comment]
This commit is before youe mention. I know it problems. I'm working on it

**Comment by aminsaedi** [comment]
a question
how can I filter this query based on `previous.base` I want `previous.base` to be `docs-src` to get list of root folders (`pwgeb` and `pingule`)
```
query MyQuery {
  allDirectory {
    edges {
      previous {
        base
      }
      node {
        relativePath
        base
        absolutePath
      }
    }
  }
}

```

**Comment by pmvald** [comment]
> a question how can I filter this query based on `previous.base` I want `previous.base` to be `docs-src` to get list of root folders (`pwgeb` and `pingule`)
> 
> ```
> query MyQuery {
>   allDirectory {
>     edges {
>       previous {
>         base
>       }
>       node {
>         relativePath
>         base
>         absolutePath
>       }
>     }
>   }
> }
> ```

For what purpose you need a list of doc names? ...

**Comment by aminsaedi** [comment]
> > a question how can I filter this query based on `previous.base` I want `previous.base` to be `docs-src` to get list of root folders (`pwgeb` and `pingule`)
> > ```
> > query MyQuery {
> >   allDirectory {
> >     edges {
> >       previous {
> >         base
> >       }
> >       node {
> >         relativePath
> >         base
> >         absolutePath
> >       }
> >     }
> >   }
> > }
> > ```
> 
> For what purpose you need a list of doc names? ...

I wanted to store tab status of each doc in the redux store.
Please look at ne commit
in this new store structure we can store state of each doc separately.
but it has an issue: each time we add new folder to `docs-src` we have to modify `doc.ts` file. (maybe it's not a problem because usually we do not add new folder to  `docs-src`)

**Comment by aminsaedi** [comment]
about the `activeDoc` in `doc` slice of store.
I added this because I need to know which doc is on screen in order to add new tab
there was two ways: 1- add new prop to `TabContainer` component and pass doc name there 2- save active doc in the store and read it from store in `tabs.tsx`
I thought the second approach is more cleaner. I can use another way if you think it's not a good idea

**Comment by pmvald** [comment]
Yes we should hardcode doc names into the store. Gatsby graph ql cannot be used for that, it is only available for generating react components at build; so you can use graph ql only inside components or hooks - not anywhere else

redux is a temporary solution - soon we will replace it with our Pinglue system which will do the same thing way easier, without all these classes and boilerplates ...

File: src/components/doc-page.tsx [hunk 3]
```diff
@@ -1,17 +1,24 @@
...
   display: "grid",
   gridTemplateColumns: "300px 1fr",
 };
 
 export default function ({ docName, title, body, slug }) {
-  const shortcodes = { Notice, Tab, Tabs };
+  const dispatch = useDispatch();
+  const shortcodes = { Notice, Tab, TabContainer };
+  useLayoutEffect(() => {
+    dispatch(changeActiveDoc(slug.split("/")[0]));
```
Comment (pmvald) [inline src/components/doc-page.tsx:20]: Make an issue: "making slug as context for doc-page component"

In addition to this task, we need slug in several other sub components (like header for generating the breadcrumb)

We will do this later

File: src/store/doc.ts [hunk 4]
```diff
@@ -1,37 +1,96 @@
...
+    (docName: DocNames, tabContainerId: string, activeTabIndex: number) =>
+        (dispatch, getState) => {
+            return dispatch(
+                activeTabChanged({ docName, tabContainerId, activeTabIndex })
+            );
+        };
+
+export const changeActiveDoc = (docName: DocNames) => (dispatch, getState) => {
+    return dispatch(activeDocChanged(docName))
+}
```
Comment (pmvald) [inline src/store/doc.ts:96]: Also count how many new lines of code we added in order to add the redux feature - later we compare it with our own tools and see simplification

**Review by pmvald — APPROVED** [review]
Looks good - please resolve the discussions and merge (you might need to do conflict resolution though)

**Comment by aminsaedi** [comment]
> Looks good - please resolve the discussions and merge (you might need to do conflict resolution though)

I don't have write access to the repo so I can't resolve file confilicts
image

**Comment by pmvald** [comment]
> > Looks good - please resolve the discussions and merge (you might need to do conflict resolution though)
> 
> I don't have write access to the repo so I can't resolve file confilicts image

Basically what you have to do first is to resolve the conflicts which is done within your own branch:

```
git fetch // update
git checkout main
git pull // update main
git checkout your-branch // go to your branch
git pull // update your branch
git merge main // merge latest changes from main
```

After this you will see some errors saying there are conflict with main - those conflicting line will be marked by a certain notation called "conflict marker" (lot of ">>>>>>>>>>>>>>>>>>") and those files containing conflict will not merge (you can run `git status` and see the unmerged files) - now editors like VSCode can recognize these conflict markers and help you resolve them by deciding what should be final value of those lines. When you resolve all conflicts save the file(s) and stage all the unmerged files:

```
git add .
```

Now commit and push to your branch - Now when you come to this page again you will see that the message "This branch has conflicts that must be resolved" on the merge button is gone and it says "no conflict" - at this stage all we need to do is to press the merge button to merge (choose the "squash and merge" option) - At this stage you may not have actually permission to do the merge. We have to check this, what I wanted to do is to give merge permission to anyone provided that the PR is approved by an admin, but I may not have done this correctly so maybe I'm the only one with merge permission. If that happens let me know and we will figure it out.

Now try to do the conflict resolution first

**Comment by pmvald** [comment]
But for this PR let me do the conflict resolution because there were some other changes in other branches and it may not be clear to you which option to choose when resolving conflicts ... but keep the previous comment in mind and at some point you will be practice it to learn how to resolve conflicts in git

**Pull request merged**